### DISCUSSION TITLE
record generic constraint

### DISCUSSION THREAD

**Discussion by ekgurney14** [dotnet/csharplang#4315] [Language Ideas]
Would it be possible in the future to allow record to be a constraint on a generic class/record.
I would like to be able to create a class where I am able to use 'with' record creation on the generic type to promote immutability.
e.g.
``
public interface IValue
{
	public int Value {get;init;}
}

public class Api<T> where T : record, IValue
{
	public T DoSomething(T item) => item with { Value = 1 };	
}

``

**Comment by huoyaoyuan** [comment]
Records are designed to not be distinguishable with normal classes. This code isn't achievable now, because the `Clone` method is not constrained.

I'd prefer to achieve things like this with HKT.

**Reply by ekgurney14** [reply]
Not super familiar with HKT.  How would this look when solving my issue?

**Reply by huoyaoyuan** [reply]
Currently, interface can only constraint on named instance members, and cannot depend on the actual type (no way to represent `typeof(this) Clone()`. HKT is designed to do more abstractions on types.

**Comment by YairHalberstadt** [comment]
See https://github.com/dotnet/csharplang/discussions/3859, https://github.com/dotnet/csharplang/issues/39#issuecomment-675463741 and https://github.com/dotnet/csharplang/discussions/3899 for earlier discussions of this

**Comment by ekgurney14** [comment]
The key takeaway from those previous discussions is that hopefully the 'with' expression will be usable with classes in c# 10.  They eliminating my need from the initial question.
